Content.Server/PneumaticCannon/PneumaticCannonSystem.cs

```diff
@@ -0,0 +1,274 @@
+using Robust.Shared.Audio;
+using Robust.Shared.IoC;
+using Robust.Shared.Localization;
+using Robust.Shared.Maths;
+using Robust.Shared.Player;
+using Robust.Shared.Random;
+
+namespace Content.Server.PneumaticCannon
+{
+    public class PneumaticCannonSystem : EntitySystem
+    {
+        [Dependency] private readonly IRobustRandom _random = default!;
+
+        private Queue<FireData> _fireQueue = new();
+        private float _accumulatedFrametime = 0f;
+        private float _dequeueInterval = .1f;
+
+        public override void Initialize()
+        {
+            base.Initialize();
+
+            SubscribeLocalEvent<PneumaticCannonComponent, ComponentInit>(OnComponentInit);
+            SubscribeLocalEvent<PneumaticCannonComponent, InteractUsingEvent>(OnInteractUsing);
+            SubscribeLocalEvent<PneumaticCannonComponent, AfterInteractEvent>(OnAfterInteract);
+        }
+
+        public override void Update(float frameTime)
+        {
+            base.Update(frameTime);
+
+            if (_fireQueue.Count == 0)
+            {
+                _accumulatedFrametime = 0;
+                return;
+            }
+
+            _accumulatedFrametime += frameTime;
+
+            if (_accumulatedFrametime >= _dequeueInterval)
+            {
+                _accumulatedFrametime -= _dequeueInterval;
+                var dat = _fireQueue.Dequeue();
+                SoundSystem.Play(Filter.Pvs(dat.User), dat.Sound.GetSound());
+                if (dat.User.TryGetComponent<CameraRecoilComponent>(out var recoil))
+                {
+                    recoil.Kick(Vector2.One * dat.Strength);
+                }
+
+                dat.Item.TryThrow(dat.Direction, dat.Strength, dat.User);
+            }
+        }
+
+        private void OnComponentInit(EntityUid uid, PneumaticCannonComponent component, ComponentInit args)
+        {
+            component.GasTankSlot = component.Owner.EnsureContainer<ContainerSlot>($"{component.Name}-gasTank");
+        }
+
+        private void OnInteractUsing(EntityUid uid, PneumaticCannonComponent component, InteractUsingEvent args)
+        {
+            args.Handled = true;
+            if (args.Used.HasComponent<GasTankComponent>() && component.GasTankSlot.CanInsert(args.Used))
+            {
+                component.GasTankSlot.Insert(args.Used);
+                args.User.PopupMessage(Loc.GetString("pneumatic-cannon-component-gas-tank-insert",
+                    ("tank", args.Used), ("cannon", component.Owner)));
+                UpdateAppearance(component);
+                return;
+            }
+
+            if (args.Used.TryGetComponent<ToolComponent>(out var tool))
+            {
+                if (tool.HasQuality(component.ModifyMode))
+                {
+                    // this is kind of ugly but it just cycles the enum
+                    var val = (byte) component.Mode;
+                    val = (byte) ((val + 1) % Enum.GetValues<PneumaticCannonFireMode>().Length);
```

Comment:
Something you can do is put a `Max` enum member at the end and then use `(int) PneumaticCannonFireMode.Max` as a way to get the length.